### PR TITLE
New Pattern - Eventbridge replay to sample and mirror prod events- with python CDK

### PR DESCRIPTION
**Pull request by sada-vela**
*Issue #, if available:*

*Description of changes:*
This pattern helps sample AWS EventBridge events (select a subset of events based on specific timeframe) and send the events from higher environments (such as production) lower environments (such as dev or stage). The events in prod represents real world data and they can be very useful in lower environments for testing new versions of applications, investigate issues or conduct performance testing. This patern will seamlessly move the prod events to lower environments in an automated and scheduled fashion. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by mavi888** [comment]
@sada-vela please follow the format for the pattern as stated in the submission: https://github.com/aws-samples/serverless-patterns/issues/new?assignees=jbesw&labels=&template=new-serverless-pattern-submission.md&title=New+pattern+submission

Cheers,

**Comment by sada-vela** [comment]
Hi @mavi888 - Here is the issue I created right after I created this PR. https://github.com/aws-samples/serverless-patterns/issues/1284

The above issue has a link to this PR. This PR contains the example-pattern.json and other files mentioned in the submission guide. Kindly take a look. Thank you!

**Comment by sada-vela** [comment]
> @sada-vela please follow the format for the pattern as stated in the submission: https://github.com/aws-samples/serverless-patterns/issues/new?assignees=jbesw&labels=&template=new-serverless-pattern-submission.md&title=New+pattern+submission
> 
> Cheers,
reply


Here is the screenshot of the issue I created along with my responses to the requirements mentioned in the submission ticket highlighted.

File: eventbridge-events-replay-cdk/python/README.md [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+## Requirements
+
+- [Create an AWS account](https://portal.aws.amazon.com/gp/aws/developer/registration/index.html) if you do not already have one and log in. The IAM user that you use must have sufficient permissions to make necessary AWS service calls and manage AWS resources.
+- [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html) installed and configured
+- [Git Installed](https://git-scm.com/book/en/v2/Getting-Started-Installing-Git)
+- [Python 3](https://www.python.org/downloads/)
+
+## Deployment Instructions
+
+# General Steps
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/README.md:23]: fix levels of headlines

File: eventbridge-events-replay-cdk/python/README.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+   ```
+   source .venv/bin/activate
+   ```
+   For a Windows platform, activate the virtualenv like this:
+   ```
+   .venv\Scripts\activate.bat
+   ```
+5. Install the Python required dependencies:
+   ```
+   pip install -r requirements.txt
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/README.md:51]: here you need to mark to get inside the python directory

File: eventbridge-events-replay-cdk/python/README.md [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+   ```
+6. Review the CloudFormation template the cdk generates for you stack using the following AWS CDK CLI command:
+   ```
+   cdk synth
+   ```
+# Deploy Lower environment stack
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/README.md:57]: fix headings

File: eventbridge-events-replay-cdk/python/cdk.json [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      "@aws-cdk/aws-ecs:arnFormatIncludesClusterName": true,
+      "@aws-cdk/core:validateSnapshotRemovalPolicy": true,
+      "@aws-cdk/aws-codepipeline:crossAccountKeyAliasStackSafeResourceName": true,
+      "@aws-cdk/aws-s3:createDefaultLoggingPolicy": true,
+      "@aws-cdk/aws-sns-subscriptions:restrictSqsDescryption": true,
+      "@aws-cdk/core:target-partitions": [
+        "aws",
+        "aws-cn"
+      ],
+      "prodAccountId": "869624704093",
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/cdk.json:37]: I hope this is not your prod account :)

File: eventbridge-events-replay-cdk/python/README.md [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+1. Open python/cdk.json file and update "prodAccountId" value to the account number of the prod account. 
+
+2. From the command line, use AWS CDK to deploy the AWS resources for the serverless application as specified in the app.py file:
+   ```
+   cdk deploy SimpleLowerEnvironmentStack --profile "stage"
+   ```
+3. Once deployment is successful, CDK will output arn of the receiving event bus. Copy this and we will use this in the next step.
+
+4. Also, this stack would have created a simple EventBridge rule and an SNS target for testing. Login to AWS console and create an email subscription for this SNS topic with your email Id so you will receive the events coming into the event bus in the stage environment to your email. More details (here) [https://docs.aws.amazon.com/sns/latest/dg/sns-email-notifications.html]
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/README.md:67]: are the email subscriptions needed for the demo, if so please add them as IaC

File: eventbridge-events-replay-cdk/python/README.md [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+# Deploy Prod environment stack
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/README.md:69]: fix headings

File: eventbridge-events-replay-cdk/python/cdk.json [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      "crossAccountEventBusArn" : "arn:aws:events:us-east-2:456818800811:event-bus/from-other-account"
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/cdk.json:38]: is this your account number?

File: eventbridge-events-replay-cdk/python/README.md [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+cdk destroy EventbridgeEventsReplayStack --profile="prod"
+```
+
+```
+cdk destroy SimpleLowerEnvironmentStack --profile="stage"
+```
+
+---
+
+Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/README.md:117]: 2023

File: eventbridge-events-replay-cdk/README.md [hunk 9]
```diff
@@ -0,0 +1,8 @@
+# Sample and Mirror Prod Events to Lower Environments
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/README.md:1]: remove this file

File: eventbridge-events-replay-cdk/python/stack.py [hunk 10]
```diff
@@ -0,0 +1,340 @@
...
+            "KinesisWriteRole",
+            assumed_by=aws_iam.ServicePrincipal(f"events.amazonaws.com"),
+            role_name="kinesis_write_role_for_events_rule",
+        )
+        kinesis_write_role.attach_inline_policy(
+            aws_iam.Policy(
+                self,
+                "write_to_kinesis_policy",
+                statements=[
+                    aws_iam.PolicyStatement(actions=["kinesis:*"], resources=["*"])
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/stack.py:61]: can you make this permission only to the kinesis stream that you are going to use?

File: eventbridge-events-replay-cdk/python/stack.py [hunk 11]
```diff
@@ -0,0 +1,340 @@
...
+                os.path.join(dirname, "functions", "replay-events")
+            ),
+            handler="replay_events.handler",
+            timeout=Duration.seconds(300),
+            environment={
+                "EVENT_BUS_NAME": main_bus.event_bus_name,
+                "EVENT_BUS_ARN": main_bus.event_bus_arn,
+                "EVENT_SOURCE_ARN": main_bus_archive.archive_arn,
+                "ALLOW_REPLAY_EVENTS_ONLY_RULE_ARN": allow_replay_only_events_rule.attr_arn,
+                "SAMPLE_START_EPOCH": "1682009582",
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/stack.py:208]: what are these dates doing here? they are hardcoded and older than today...

File: eventbridge-events-replay-cdk/python/stack.py [hunk 12]
```diff
@@ -0,0 +1,340 @@
...
+            timeout=Duration.seconds(300),
+        )
+        describe_replay_lambda.role.attach_inline_policy(
+            aws_iam.Policy(
+                self,
+                "DescribeReplayPolicy",
+                statements=[
+                    aws_iam.PolicyStatement(
+                        actions=["events:DescribeReplay"],
+                        resources=["*"],
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/stack.py:241]: can you limit this to the event bus?

File: eventbridge-events-replay-cdk/python/simple_lower_env_stack.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+            "SnsWriteRole",
+            assumed_by=aws_iam.ServicePrincipal(f"events.amazonaws.com"),
+            role_name="sns_write_role",
+        )
+        sns_write_role.attach_inline_policy(
+            aws_iam.Policy(
+                self,
+                "write_to_sns",
+                statements=[
+                    aws_iam.PolicyStatement(actions=["sns:*"], resources=["*"])
```
Comment (mavi888) [inline eventbridge-events-replay-cdk/python/simple_lower_env_stack.py:46]: can you give the right permissions here?

**Comment by mavi888** [comment]
Also there is an image in the readme but it is hard to see. Can you make it bigger, and also in the READMe you might want to talk about the lambda functions you have in place.

This pattern is quite big, if you think this can be a repo instead a pattern, submit this as a repo instead (https://serverlessland.com/repos)

**Comment by mavi888** [comment]
@sada-vela I will close this PR by the end of the week if the comments don't get addressed. 
Cheers

**Pull request closed without merging**